run-build.sh

```diff
@@ -0,0 +1,183 @@
+            shift
+            ;;
+        --runtime-id)
+            CUSTOM_BUILD_ARGS="/p:Rid=\"$2\""
+            args=( "${args[@]/$1}" )
+            args=( "${args[@]/$2}" )
+            shift
+            ;;
+        # This is here just to eat away this parameter because CI still passes this in.
+        --targets)            
+            args=( "${args[@]/$1}" )
+            args=( "${args[@]/$2}" )
+            shift
+            ;;
+        --linux-portable)
+            LINUX_PORTABLE_INSTALL_ARGS="--runtime-id linux-x64"
+            CUSTOM_BUILD_ARGS="/p:Rid=\"linux-x64\" /p:OSName=\"linux\" /p:IslinuxPortable=\"true\""
+            args=( "${args[@]/$1}" )
+            ;;
+        --help)
+            echo "Usage: $0 [--configuration <CONFIGURATION>] [--targets <TARGETS...>] [--skip-prereqs] [--nopackage] [--docker <IMAGENAME>] [--help]"
+            echo ""
+            echo "Options:"
+            echo "  --configuration <CONFIGURATION>     Build the specified Configuration (Debug or Release, default: Debug)"
+            echo "  --skip-prereqs                      Skip checks for pre-reqs in dotnet_install"
+            echo "  --nopackage                         Skip packaging targets"
+            echo "  --nobuild                           Skip building, showing the command that would be used to build"
+            echo "  --docker <IMAGENAME>                Build in Docker using the Dockerfile located in scripts/docker/IMAGENAME"
+            echo "  --linux-portable                    Builds the Linux portable .NET Tools instead of a distro-specific version."
+            echo "  --help                              Display this help message"
+            exit 0
+            ;;
+        *)
+            break
+            ;;
+    esac
+
+    shift
+done
+
+# $args array may have empty elements in it.
+# The easiest way to remove them is to cast to string and back to array.
+# This will actually break quoted arguments, arguments like
+# -test "hello world" will be broken into three arguments instead of two, as it should.
+temp="${args[@]}"
+args=($temp)
+
+# Create an install directory for the stage 0 CLI
+[ -z "$DOTNET_INSTALL_DIR" ] && export DOTNET_INSTALL_DIR=$REPOROOT/.dotnet_stage0/$ARCHITECTURE
+[ -d "$DOTNET_INSTALL_DIR" ] || mkdir -p $DOTNET_INSTALL_DIR
+
+# We also need to pull down a project.json based CLI that is used by some tests
+# so create another directory for that.
+[ -z "$DOTNET_INSTALL_DIR_PJ" ] && export DOTNET_INSTALL_DIR_PJ=$REPOROOT/.dotnet_stage0PJ/$ARCHITECTURE
+[ -d "$DOTNET_INSTALL_DIR_PJ" ] || mkdir -p $DOTNET_INSTALL_DIR_PJ
+
+export DOTNET_SKIP_FIRST_TIME_EXPERIENCE=1
+
+# Don't resolve shared frameworks from user or global locations
+export DOTNET_MULTILEVEL_LOOKUP=0
+
+# Install a stage 0
+(set -x ; "$REPOROOT/scripts/obtain/dotnet-install.sh" --channel "release/2.0.2" --install-dir "$DOTNET_INSTALL_DIR" --architecture "$ARCHITECTURE" $LINUX_PORTABLE_INSTALL_ARGS)
```

Comment:
I understand you copied all this from the dotnet cli repo, and if it *just works*, then it's OK as a starting point (EDIT: now looking at the amount of scripts you added, I'm not that comfortable either, I thought it was just this script). The reality is that they need to be able to build in several different environments, that's why the build scripts and dockerfiles are so complex and install everything from scratch.
Our complexity is orders of magnitude less, our build could just be to run a container with an already set up image such as this one https://hub.docker.com/r/microsoft/dotnet/ and call `dotnet build` from it, reducing these amount of scripts to less than 10 lines (probable less than 3 lines)

Reply:
And of course that has the added benefit that we talked about, which is that the first time pulling the image might take long, but subsequent would be immediate as they use the cached image. Where as these scripts install everything from scratch every single time, regardless of cache